### PR TITLE
[onnx.export] Avoid building vals_to_params_map

### PR DESCRIPTION
**Pull request by gustavla**
This PR is part of an effort to speed up torch.onnx.export (#121422).

- Building vals_to_params_map costs linear time in N (number of nodes), when instead we can index into this dictionary directly.
- No need to call HasField on the final else, since c10::nullopt is the default returned value if a field does not exist.
- Resolves (3) in #121422.

### CODE REVIEW

**Comment by gustavla** [comment]
Is it possible that the failing test is a non-deterministic test that happens to trigger? It's hard to see how these changes can cause a regression. I will try to see if I can run this test locally.

**Comment by gustavla** [comment]
In fact, CI is even pointing this out:
> UNSTABLE - The following job failed but was likely due to flakiness present on trunk and has been marked as unstable:

**Review by BowenBao — APPROVED** [review]
Thanks for contribution, LGTM!

**Review by thiagocrepaldi — APPROVED** [review]

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 1]
```diff
@@ -295,10 +295,8 @@ Node* CloneNodeToGraph(
     std::shared_ptr<Graph> n_graph,
     const ParamMap& params_dict,
     int opset_version) {
-  auto vals_to_params_map =
-      buildValueToParamsMap(n->owningGraph()->block(), params_dict);
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:299]: I realize this is called for each node, so indeed this is O(N).

**Pull request closed without merging**